### DISCUSSION TITLE
Scaling recommended method change implications and Colyseus cloud

### DISCUSSION THREAD

**Discussion by rajdev-cuemath** [colyseus/colyseus#582] [General]
Hi everyone,

I see that from v0.15, the recommended method of scaling has changed. This leads me to having multiple questions.

1) Can we still use the previous dynamic proxy method to scale? If yes, why what are the tradeoffs and what are the reasons behind the change in recommended method?

2) What issues, if any we may face with the previous method while using 0.15 or above?

3) Is it advisable to use an older version of colyseus along with dynamic proxy?

4) I see that we may be able to use colyseus proxy from mid June. What cost should I expect if I want to create a small gaming app with approx 5k daily active users during peak evening hours. Let's assume the game is chess.

5) I am planning to create an express integrated colyseus. It looks like options.server will also be deprecated, does it mean I will not be able to use an express server for my app?

 Can someone please help me with these?
 Thanks

**Comment by endel** [comment]
Hi @rajdev-cuemath, 

> Can we still use the previous dynamic proxy method to scale? If yes, why what are the tradeoffs and what are the reasons behind the change in recommended method?

Sure, you can still use the proxy. The upsides of the newer method are:

- Easier to configure and understand
- More flexible to deploy
- Traffic between client and server is **direct** (a couple of ms faster than with the proxy)

> What issues, if any we may face with the previous method while using 0.15 or above?

There are 2 major downsides of the proxy: 
1. single point of failure: if the proxy goes down for some reason, **every** client connection will be dropped.
2. additional redis state persistence which is prone to failure: during ungraceful shutdowns, a Colyseus node may not unregister itself from Redis, causing the proxy to still try to use it as one of its destinations - and causing `"seat reservation"` errors. 

> Is it advisable to use an older version of colyseus along with dynamic proxy?

No problem using the latest version of the proxy with an older version of the framework. The proxy should be compatible with 0.13, 0.14, and 0.15 of the server.

> I see that we may be able to use colyseus proxy from mid June. What cost should I expect if I want to create a small gaming app with approx 5k daily active users during peak evening hours. Let's assume the game is chess.

5k daily active users are not equal to 5k concurrent users (CCU). For a game like chess, a single process on a cheap VM may easily handle 2k CCU.

> I am planning to create an express integrated colyseus. It looks like options.server will also be deprecated, does it mean I will not be able to use an express server for my app?

Transport options have been refactored, and `server` is now an option for `WebSocketTransport`: https://docs.colyseus.io/server/transport/#default-websocket-transport-via-ws

```typescript
const app = express();
const server = http.createServer(app);
const gameServer = new Server({
    transport: new WebSocketTransport({ server })
})
```

**Reply by rajdev-cuemath** [reply]
@endel Thanks a lot for the quick response, it is really helpful.
